cost/aws/elb/clb_unused/AWS_Delete_Unused_CLB.pt

```diff
@@ -0,0 +1,242 @@
+    field "description", jmes_path(col_item,"Description")
+    field "state", jmes_path(col_item,"State")
+    field "instanceId", jmes_path(col_item,"InstanceId")
+    field "loadBalancerName", val(iter_item, "loadBalancerName")
+    end
+  end
+end
+
+#https://www.cloudconformity.com/conformity-rules/ELB/unused-elastic-load-balancers.html
+script "js_aws_clb_filter_map", type: "javascript" do
+  parameters "ds_aws_clb_list_tags", "ds_aws_instances_health", "param_exclude_tags"
+
+  result "content"
+  code <<-EOS
+  var content=[]
+  for(var i=0; i<ds_aws_clb_list_tags.length; i++){
+    clb=ds_aws_clb_list_tags[i]
+
+    // Constructing availabilityZones with comma separated to display in detail_template
+    var availabilityZones=clb['availabilityZones']
+    var s_availabilityZones=""
+    for(var j=0; j < availabilityZones.length; j++){
+      availabilityZone = availabilityZones[j]
+      s_availabilityZones = s_availabilityZones + ', '+ availabilityZone
+    }
+
+    // Constructing IntanceIds with comma separated to display in detail_template
+    var instancesList = clb['instancesList']
+    var s_instanceIds=""
+    for(var j=0; j < instancesList.length; j++){
+      instanceId = instancesList[j]
+      s_instanceIds = s_instanceIds + ', '+ instanceId
+    }
+
+    //Constructing Tags of individual CLB into key=value format. Check, if the tag present in entered param_exclude_tags, ignore the CLB if the tag matches/present.
+    var tags = clb['tags']
+    var isTagMatched=false
+    var tagKeyValue=""
+    for(var j=0; j < tags.length; j++){
+      tag = tags[j]
+      if(param_exclude_tags.indexOf(tag['tagKey']+'='+tag['tagValue']) !== -1){
+        isTagMatched = true;
+      }
+      // Constructing tags with comma separated to display in detail_template
+      tagKeyValue = tagKeyValue + ', '+ tag['tagKey']+'='+tag['tagValue']
+    }
+
+    //If the CLB tag does not match with entered param_exclude_tags, then check if any instance exists in the CLB and check for instances health.
+    if(!(isTagMatched)){
+      clb_details={
+        loadBalancerName: clb['loadBalancerName'],
+        instanceId:(s_instanceIds.slice(2)),
+        availabilityZones:(s_availabilityZones.slice(2)),
+        tagKeyValue:(tagKeyValue.slice(2))
+      }
+      if(instancesList.length > 0){
+        var isInstanceNotHealthy=false
+        // Check if the CLB has no healthy instance with state and description matches for the values mentioned below
+    var state= "OutOfService"
+        var description= "Instance has failed at least the UnhealthyThreshold number of health checks consecutively."
+        for(var k=0; k < ds_aws_instances_health.length; k++){
+          instance= ds_aws_instances_health[k]
+          if(((instance['loadBalancerName']) == (clb['loadBalancerName'])) && (instance['state'] == state && instance['description'] == description)){
```

Comment:
Why do you need to validate both instance state and description here?

Reply:
We can have CLB with same state (OutOfService) and different descriptions like “1. Instance registration is still in progress,2.Instance is in stopped state”
![image](https://user-images.githubusercontent.com/47382786/55232315-1e8b9600-524b-11e9-99d3-ee6ac7a282fc.png)

Where instance is healthy with state:OutofService
Hence we are checking both state (OutofService) and description(Instance has failed at least the UnhealthyThreshold number of health checks consecutively.) to identify unhealthy instance
AWS API link: https://docs.aws.amazon.com/elasticloadbalancing/2012-06-01/APIReference/API_DescribeInstanceHealth.html for instance has different states and description
We referred: https://www.cloudconformity.com/conformity-rules/ELB/unused-elastic-load-balancers.html as mentioned in the requirement to determine unhealthy instance.

Reply:
@avinashbhashyam-rs If you are satisfied with the response can you please resolve the conversation.

Reply:
@avinashbhashyam-rs If you are satisfied with the response, can you please resolve the conversation.